### PR TITLE
Add fast path for linq count with predicate

### LINKED ISSUE #102696: LINQ - use TryGetSpan in Enumerable.Count with predicate
### Description

similar how [Max.cs](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Linq/src/System/Linq/Max.cs) & [Min](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Linq/src/System/Linq/Min.cs) try to convert to span

https://github.com/dotnet/runtime/blob/35e4aad602ee3c28330e94746cfd4e0d4569b66f/src/libraries/System.Linq/src/System/Linq/Max.cs#L107
https://github.com/dotnet/runtime/blob/35e4aad602ee3c28330e94746cfd4e0d4569b66f/src/libraries/System.Linq/src/System/Linq/Min.cs#L210

Iterating over a span when possible would be faster then iterating over enumerable.

### Configuration

// * Summary *

BenchmarkDotNet v0.13.12, Windows 11 (10.0.22631.3593/23H2/2023Update/SunValley3)
11th Gen Intel Core i9-11950H 2.60GHz, 1 CPU, 16 logical and 8 physical cores
.NET SDK 8.0.205
  [Host]   : .NET 8.0.5 (8.0.524.21615), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI
  ShortRun : .NET 8.0.5 (8.0.524.21615), X64 RyuJIT AVX-512F+CD+BW+DQ+VL+VBMI

Job=ShortRun  IterationCount=3  LaunchCount=1
WarmupCount=3

| Method                | count   | Mean            | Error           | StdDev        | Allocated |
|---------------------- |-------- |----------------:|----------------:|--------------:|----------:|
| Count_IEnumerable     | 100     |      1,732.7 ns |        741.6 ns |      40.65 ns |      48 B |
| Count_List            | 100     |        209.5 ns |        427.7 ns |      23.44 ns |      40 B |
| Count_Array           | 100     |        196.8 ns |        256.8 ns |      14.08 ns |      32 B |
| Count_IEnumerable_New | 100     |      1,638.3 ns |      1,091.5 ns |      59.83 ns |      48 B |
| Count_List_New        | 100     |        164.6 ns |        366.9 ns |      20.11 ns |         - |
| Count_Array_New       | 100     |        114.8 ns |        135.1 ns |       7.41 ns |         - |
| Count_IEnumerable     | 10000   |    161,695.6 ns |     64,981.1 ns |   3,561.83 ns |      48 B |
| Count_List            | 10000   |     42,592.3 ns |     17,101.9 ns |     937.41 ns |      40 B |
| Count_Array           | 10000   |     44,711.1 ns |     15,163.0 ns |     831.14 ns |      32 B |
| Count_IEnumerable_New | 10000   |    161,839.9 ns |    210,702.2 ns |  11,549.30 ns |      48 B |
| Count_List_New        | 10000   |     36,906.0 ns |     32,625.2 ns |   1,788.30 ns |         - |
| Count_Array_New       | 10000   |     36,233.2 ns |      7,040.9 ns |     385.94 ns |         - |
| Count_IEnumerable     | 1000000 | 14,854,601.0 ns | 12,265,409.7 ns | 672,308.49 ns |      60 B |
| Count_List            | 1000000 |  6,323,082.3 ns |  9,923,756.9 ns | 543,954.60 ns |      43 B |
| Count_Array           | 1000000 |  6,402,595.8 ns |  2,037,387.6 ns | 111,676.09 ns |      35 B |
| Count_IEnumerable_New | 1000000 | 15,711,805.2 ns | 13,675,725.7 ns | 749,612.67 ns |      54 B |
| Count_List_New        | 1000000 |  4,253,217.7 ns |  3,217,607.6 ns | 176,367.93 ns |       3 B |
| Count_Array_New       | 1000000 |  4,241,221.4 ns |  4,164,944.1 ns | 228,294.64 ns |       3 B |

// * Legends *
  count     : Value of the 'count' parameter
  Mean      : Arithmetic mean of all measurements
  Error     : Half of 99.9% confidence interval
  StdDev    : Standard deviation of all measurements
  Allocated : Allocated memory per single operation (managed only, inclusive, 1KB = 1024B)
  1 ns      : 1 Nanosecond (0.000000001 sec)

// * Diagnostic Output - MemoryDiagnoser *

// ***** BenchmarkRunner: End *****
Run time: 00:01:54 (114.04 sec), executed benchmarks: 18

Global total time: 00:02:04 (124.39 sec), executed benchmarks: 18

<details>

```cs
using System.Runtime.InteropServices;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

namespace MyBenchmarks;

[MemoryDiagnoser(false)]
[ShortRunJob]
public class EnumerableExcept
{
    [Params(100, 10000, 1000000)]
    public int count;

    private IEnumerable<int> collection;
    
    private List<int> list;
    private int[] array;


    [GlobalSetup]
    public void GlobalSetup()
    {
        Random random = new Random(42);
        collection = Enumerable.Range(0, count).Select(i => random.Next());
        
        list = collection.ToList();
        array = collection.ToArray();
    }

    [Benchmark]
    public int Count_IEnumerable() => collection.Count(x => x % 2 == 0);

    [Benchmark]
    public int Count_List() => list.Count(x => x % 2 == 0);

    [Benchmark]
    public int Count_Array() => array.Count(x => x % 2 == 0);

    [Benchmark]
    public int Count_IEnumerable_New() => MyEnumerable.Count(collection, x => x % 2 == 0);

    [Benchmark]
    public int Count_List_New() => MyEnumerable.Count(list, x => x % 2 == 0);

    [Benchmark]
    public int Count_Array_New() => MyEnumerable.Count(array, x => x % 2 == 0);
}

public static class MyEnumerable
{

    public static int Count<TSource>(IEnumerable<TSource> collection, Func<TSource, bool> predicate)
    {
        ArgumentNullException.ThrowIfNull(collection);
        ArgumentNullException.ThrowIfNull(predicate);

        if (collection is TSource[] array) 
        {
            return SpanCount(array, predicate);
        }

        if (collection is List<TSource> list)
        {
            return SpanCount(CollectionsMarshal.AsSpan(list), predicate);
        }

        int count = 0;
        foreach (var item in collection)
        {
            if (predicate(item))
            {
                count++;
            }
        }

        return count;

        static int SpanCount(Span<TSource> span, Func<TSource, bool> predicate)
        {
            int count = 0;
            for (int i = 0; i < span.Length; i++)
            {
                if (predicate(span[i]))
                {
                    count++;
                }
            }

            return count;
        }
    }
}

public class Program
{
    public static void Main(string[] args)
    {
        var summary = BenchmarkRunner.Run<EnumerableExcept>();
    }
}

```

</details>

### Regression?

No

### Analysis

https://github.com/dotnet/runtime/blob/35e4aad602ee3c28330e94746cfd4e0d4569b66f/src/libraries/System.Linq/src/System/Linq/Count.cs#L62 call `source.TryGetSpan(out ReadOnlySpan<TSource> span)` and then a dedicated foreach loop for the read-only-span

### PR DESCRIPTION
**Pull request by neon-sunset**
Fixes https://github.com/dotnet/runtime/issues/102696

A simple change to take fast path when the source is span-able. Also move non-span loop to a local function to make it easier for the JIT to inline the method alongside the lambda, when available, per callsite, avoiding a single lambda only devirtualization due to the loop not being inlined.

### CODE REVIEW

**Comment by neon-sunset** [comment]
@EgorBot -arm64 -intel
```cs
using BenchmarkDotNet.Attributes;
using System.Buffers;
using BenchmarkDotNet.Running;

BenchmarkRunner.Run<Perf_Count>(args: args);

[MemoryDiagnoser]
[DisassemblyDiagnoser(maxDepth: 4)]
public class Perf_Count
{
    [Params(1, 5, 10, 100, 10_000)]
    public int Length;

    int[]? _array;
    IEnumerable<int>? _select;

    [GlobalSetup]
    public void Setup()
    {
        _array = Enumerable.Range(0, Length).ToArray();
        _select = _array.Select(i => i);
    }

    [Benchmark]
    public int Array() => _array!.Count(i => i % 2 == 0);

    [Benchmark]
    public int Select() => _select!.Count(i => i % 2 == 0);
}
```

File: src/libraries/System.Linq/src/System/Linq/Count.cs [hunk 1]
```diff
@@ -59,19 +60,45 @@ public static int Count<TSource>(this IEnumerable<TSource> source, Func<TSource,
...
+
+            [StackTraceHidden]
+            static int CountSpan(ReadOnlySpan<TSource> source, Func<TSource, bool> predicate)
             {
-                checked
+                int count = 0;
+                foreach (TSource element in source)
                 {
-                    if (predicate(element))
+                    checked
```
Comment (jkotas) [inline src/libraries/System.Linq/src/System/Linq/Count.cs:73]: You do not need `checked` here. The Span can only have so many elements.

**Comment by jkotas** [comment]
> Also move non-span loop to a local function to make it easier for the JIT to inline the method alongside the lambda, when available, per callsite, avoiding a single lambda only devirtualization due to the loop not being inlined.

This split adds code bloat. We do not have an artificial split like this in other places that use `TryGetSpan`. Is it really worth it here?

File: src/libraries/System.Linq/src/System/Linq/Count.cs (hunk 1 above)
Comment (neon-sunset) [reply src/libraries/System.Linq/src/System/Linq/Count.cs:73]: Thanks.

**Comment by neon-sunset** [comment]
> This split adds code bloat. We do not have an artificial split like this in other places that use `TryGetSpan`. Is it really worth it here?

The goal behind this is to allow the JIT to inline the `.Count` call (which it does as of this PR) and then also optimize away type handle comparisons in `.TryGetSpan` (which it currently doesn't, at least for the instance fields in the benchmark). Without hoisting enumerable count to a local function or a method this is not possible as `foreach` brings EH which blocks the inlining. When inlining is blocked, only a single delegate will be devirtualized in the span path given current DPGO restrictions if I understand it correctly.

This also allows the JIT to optimize away the precondition checks when possible (almost noise given enumerator allocation but hey).

Second variant which moves the span loop to a local function shaves off ~100B of codegen impact size on enumerable path. For the span path, this PR is codegen size win.

**Comment by EgorBo** [comment]
> ❌ Benchmark failed on Intel

Sadly, it seems to be some general unrelated BDN issue - it just dies sometimes (I think Stephen hit it too once). Looks like it's `DisassemblyDiagnoser` specific.

**Comment by neon-sunset** [comment]
@jkotas After reviewing ARM64 results and their disassembly, it appears that my assumption on per-callsite delegate inling was proven wrong - I *think* JIT gathers the profile in the instrumented Count/CountSpan compilations before they get inlined, causing only a single instance of the delegate to get inlined in both loop bodies (winning non-deterministically?):

```asm
M00_L02: ;; first loop start
       ldr       w1, [x23, w22, uxtw #2]
       ldr       x0, [x21, #0x18]
       add       x2, x26, #0x18
       cmp       x0, x2
       b.ne      M00_L03
       tst       w1, #7 ;; i % 8 is 0 instead of i % 2 is 0
       b.ne      M00_L05
       b         M00_L04
M00_L03:
       ldr       x0, [x21, #8]
       ldr       x2, [x21, #0x18]
       blr       x2
       cbz       w0, M00_L05
M00_L04:
       add       w25, w25, #1
M00_L05:
       add       w22, w22, #1
       cmp       w22, w24
       b.lt      M00_L02
M00_L06:
       ldr       x22, [x20, #8]
       cbz       x22, M00_L18
M00_L07:
       cbz       x22, M00_L24
       movz      w23, #0x1
       ldr       x1, [x19]
       movz      x0, #0x61d0
       movk      x0, #0x8d05, lsl #16
       movk      x0, #0xe8f5, lsl #32
       cmp       x1, x0
       b.ne      M00_L19
       add       x20, x19, #0x10
       ldr       w21, [x19, #8]
M00_L08:
       cbz       w23, M00_L23
       mov       w24, wzr
       mov       w19, wzr
       cmp       w21, #0
       b.le      M00_L13
       movz      x26, #0xb360
       movk      x26, #0x8d61, lsl #16
       movk      x26, #0xe8f5, lsl #32
M00_L09:
       ldr       w1, [x20, w19, uxtw #2]
       ldr       x0, [x22, #0x18]
       add       x2, x26, #0x18
       cmp       x0, x2
       b.ne      M00_L10
       tst       w1, #7 ;; same i % 8 is 0, as expected
       b.ne      M00_L12
       b         M00_L11
M00_L10:
       ldr       x0, [x22, #8]
       ldr       x2, [x22, #0x18]
       blr       x2
       cbz       w0, M00_L12
M00_L11:
       add       w24, w24, #1
M00_L12:
       add       w19, w19, #1
       cmp       w19, w21
       b.lt      M00_L09
```

Given the above, please let me know what you think is the appropriate course of action:

- Keep current variant
- Outline span loop (mitigates DPGO impact on enumerable in bi/polymorphic scenarios where span was favoured, but prevents method inlining which has both pros and cons, more cons if JIT learns to optimize away TryGetSpan)
- Remove outlining for both span and enumerable, enumerable will be impacted in bi/polymorphic scenarios but it will be more consistent with other LINQ code

**Comment by neon-sunset** [comment]
> My preference is to keep all Linq code consistent.

Personally, I think all top-level LINQ methods that have TryGetSpan + EH from IEnumerable<T> path in them are making a suboptimal implementation choice but that's not for me to decide.

File: src/libraries/System.Linq/src/System/Linq/Count.cs [hunk 2]
```diff
@@ -3,6 +3,7 @@
 
 using System.Collections;
 using System.Collections.Generic;
+using System.Diagnostics;
```
Comment (jkotas) [inline src/libraries/System.Linq/src/System/Linq/Count.cs:6]:
```suggestion
```
?

**Comment by neon-sunset** [comment]
@jkotas do you mind if I update this PR to include a similar optimization for `.Any` and `.All` predicate overloads? Both are very popular for scanning arrays/lists.

**Comment by stephentoub** [comment]
As noted in https://github.com/dotnet/runtime/issues/102696#issuecomment-2132227035, https://github.com/dotnet/runtime/compare/main...stephentoub:runtime:fasterlinqenumeration highlights that it's not just Count/Any/All, but many more APIs that would need to be updated to be made consistent. I've not submitted a PR for that yet though because of the regressions that it brings in addition to the improvements. Most of those were from special-casing IList as well, but even for the checks for array/list that the span checks incur, it adds some overhead that will show up if these APIs are used on non-arrays/lists frequently.

**Comment by neon-sunset** [comment]
Fair enough, but on the other hand this change (specifically, the variant that moves EH blocking inlining to local function) replaces the pattern that is difficult for JIT to optimize to the one that is easier to optimize and that has higher likelihood of improving (I find it surprising it still does not optimize away arr.GetType() != typeof(int[]) given it's a direct scalar comparison).

It pays with about two branches for setup to iterate directly on the span. It's cheaper than `is IList<T>` -> `.TryGetSpan` and I feel like such kind of small changes are easier to reason about and introduce vs larger rewrites.

**Comment by neon-sunset** [comment]
Please let me know what are the next steps with this change if any, thanks!

**Comment by stephentoub** [comment]
I added in use of TryGetSpan to the other relevant sinks.

**Review by stephentoub — APPROVED** [review]
Thanks.

**Pull request merged**